src/connector/src/sink/snowflake.rs

```diff
@@ -236,31 +227,87 @@ impl SnowflakeSinkWriter {
                 TimestamptzHandlingMode::UtcString,
                 TimeHandlingMode::String,
             ),
-            row_counter: 0,
-            payload: String::new(),
-            max_batch_row_num,
             // initial value of `epoch` will start from 0
             epoch: 0,
-        }
+            // will be initialized after the begin of each epoch
+            streaming_uploader: None,
+            file_suffix: None,
+            has_data: false,
+        })
     }
 
-    /// reset the `payload` and `row_counter`.
-    /// shall *only* be called after a successful sink to s3.
-    fn reset(&mut self) {
-        self.payload.clear();
-        self.row_counter = 0;
+    /// update the streaming uploader as well as the file suffix.
+    /// note: should *only* be called when a new epoch begins.
+    async fn update_streaming_uploader(&mut self) -> Result<()> {
+        self.file_suffix = Some(self.file_suffix());
+        let path =
+            generate_s3_file_name(self.s3_client.s3_path(), self.file_suffix.as_ref().unwrap());
+        let uploader = self
+            .s3_client
+            .opendal_s3_engine
+            .streaming_upload(&path)
+            .await
+            .map_err(|err| {
+                SinkError::Snowflake(format!(
+                    "failed to create the streaming uploader of opendal s3 engine for epoch {}, error: {}",
+                    self.epoch,
+                    err
+                ))
+            })?;
+        self.streaming_uploader = Some(uploader);
+        // we don't have data at the beginning of each epoch
+        self.has_data = false;
+        Ok(())
     }
 
-    fn at_sink_threshold(&self) -> bool {
-        self.row_counter >= self.max_batch_row_num
+    /// write data to the current streaming uploader for this epoch.
+    async fn streaming_upload(&mut self, data: Bytes) -> Result<()> {
```

Comment:
I think we can lazily initialize the uploader when there is new non-empty data coming, and then we can remove the `has_data` flag. The `Option` of streaming uploader is the flag to specify whether there is any data written.